snap/gadget.go

```diff
@@ -20,614 +20,24 @@
-	Image string `yaml:"image"`
-	// Offset the image is written at
-	Offset GadgetSize `yaml:"offset"`
-	// OffsetWrite describes a 32-bit address, within the volume, at which
-	// the offset of current image will be written. The position may be
-	// specified as a byte offset relative to the start of a named structure
-	OffsetWrite GadgetRelativeOffset `yaml:"offset-write"`
-	// Size of the image, when empty size is calculated by looking at the
-	// image
-	Size GadgetSize `yaml:"size"`
-
-	Unpack bool `yaml:"unpack"`
-}
-
-type VolumeUpdate struct {
-	Edition  editionNumber `yaml:"edition"`
-	Preserve []string      `yaml:"preserve"`
-}
-
-// GadgetConnect describes an interface connection requested by the gadget
-// between seeded snaps. The syntax is of a mapping like:
-//
-//  plug: (<plug-snap-id>|system):plug
-//  [slot: (<slot-snap-id>|system):slot]
-//
-// "system" indicates a system plug or slot.
-// Fully omitting the slot part indicates a system slot with the same name
-// as the plug.
-type GadgetConnection struct {
-	Plug GadgetConnectionPlug `yaml:"plug"`
-	Slot GadgetConnectionSlot `yaml:"slot"`
-}
-
-type GadgetConnectionPlug struct {
-	SnapID string
-	Plug   string
-}
-
-func (gcplug *GadgetConnectionPlug) Empty() bool {
-	return gcplug.SnapID == "" && gcplug.Plug == ""
-}
-
-func (gcplug *GadgetConnectionPlug) UnmarshalYAML(unmarshal func(interface{}) error) error {
-	var s string
-	if err := unmarshal(&s); err != nil {
-		return err
-	}
-	snapID, name, err := parseSnapIDColonName(s)
-	if err != nil {
-		return fmt.Errorf("in gadget connection plug: %v", err)
-	}
-	gcplug.SnapID = snapID
-	gcplug.Plug = name
-	return nil
-}
-
-type GadgetConnectionSlot struct {
-	SnapID string
-	Slot   string
-}
-
-func (gcslot *GadgetConnectionSlot) Empty() bool {
-	return gcslot.SnapID == "" && gcslot.Slot == ""
-}
-
-func (gcslot *GadgetConnectionSlot) UnmarshalYAML(unmarshal func(interface{}) error) error {
-	var s string
-	if err := unmarshal(&s); err != nil {
-		return err
-	}
-	snapID, name, err := parseSnapIDColonName(s)
-	if err != nil {
-		return fmt.Errorf("in gadget connection slot: %v", err)
-	}
-	gcslot.SnapID = snapID
-	gcslot.Slot = name
-	return nil
-}
-
-func parseSnapIDColonName(s string) (snapID, name string, err error) {
-	parts := strings.Split(s, ":")
-	if len(parts) == 2 {
-		snapID = parts[0]
-		name = parts[1]
-	}
-	if snapID == "" || name == "" {
-		return "", "", fmt.Errorf(`expected "(<snap-id>|system):name" not %q`, s)
-	}
-	return snapID, name, nil
-}
-
-func systemOrSnapID(s string) bool {
-	if s != "system" && len(s) != validSnapIDLength {
-		return false
-	}
-	return true
-}
-
 // ReadGadgetInfo reads the gadget specific metadata from gadget.yaml
 // in the snap. classic set to true means classic rules apply,
 // i.e. content/presence of gadget.yaml is fully optional.
-func ReadGadgetInfo(info *Info, classic bool) (*GadgetInfo, error) {
+func ReadGadgetInfo(info *Info, classic bool) (*gadget.Info, error) {
```

Comment:
Do we want/need this? Or we will get rid of this eventually? Feels slightly redundant to have "snap.ReadGadgetInfo()" and "gadget.ReadGadgetInfo()" but then one takes "snap.Info" the other just a dir. No strong opinion, just curious.

Reply:
I think it may be useful when we reach the stage when we'll have a command to assemble the image. At this point we'll either drop the redundant piece or refactor it. No strong feelings though.